build.py

```diff
@@ -1,160 +1,127 @@
+    parser.add_argument("--mem-conf", type=Path,
+                        help=Help.MEM_CONF)
     parser.add_argument("--output", "-o", type=Path,
                         help=Help.OUTPUT)
     args = parser.parse_args(argv[1:])
-    assert not args.core in args.corunner_id
+    assert args.core not in cor_cores(args.corunner)
     if args.output is None:
         args.output = args.build_dir / "program.elf"
     return args
 
-
-def write_template(output_filename, template, context):
-    output_filename.parent.mkdir(exist_ok=True, parents=True)
-    with open(output_filename, "w") as fileh:
-        fileh.write(Template(template).substitute(context))
-
-
 def gen_agent_config(output_filename, name, core):
     write_template(output_filename, AGENT_CONFIG_HJSON_TEMPLATE, {
         "agent_name": name,
         "agent_core": core,
     })
 
-
-def gen_corunner_config(output_filename, identifier, symbol, object_file):
-    write_template(output_filename, CORUNNER_CONFIG_HJSON_TEMPLATE, {
+def gen_corunner_config(conf_filename, identifier, symbol, object_file, kmem_filename):
+    write_template(conf_filename, CORUNNER_CONFIG_HJSON_TEMPLATE, {
         "corunner_id": identifier,
         "corunner_symbol": symbol,
         "corunner_object": str(object_file)
     })
+    write_template(kmem_filename, CORUNNER_KMEMORY_JSON_TEMPLATE, {
+        'symbol': symbol,
+    })
 
-
-def gen_corunner_source(output_filename, symbol, *, sram=False):
+def gen_corunner_source(output_filename, symbol, sram=dict()):
     cmd = [sys.executable, TOP_DIR / "scripts" / "gen-corunner.py", symbol]
     if sram:
         cmd += ["--sram"]
+        if 'nop' in sram.keys():
+          cmd += ["--nop", str(sram['nop'])]
+        if 'start' in sram.keys():
+          cmd += ["--startaddr", str(sram['start'])]
+        if 'size' in sram.keys():
+          cmd += ["--tablesize", str(sram['size'])]
+        if 'stride' in sram.keys():
+          cmd += ["--stride", str(sram['stride'])]
     else:
         cmd += ["--jump", "2048"]
     with subprocess.Popen(cmd, stdout=subprocess.PIPE, universal_newlines=True) as proc:
         with open(output_filename, "w") as fileh:
             fileh.write(proc.stdout.read())
 
+def gen_kmem_final(default, config, memreport, kdbv, tasks, corunners=list()):
+    config_json = load_json(config)
+    cmd = [sys.executable, TOP_DIR / 'scripts' / 'gen-kmem.py', '--config', config]
+    del_list = []
+    config_json['memreport'] = str(memreport)
+    config_json['default_kmemory'] = str(default)
+    for el in config_json['elements']:
+        if el['type'] == 'corunner':
+            if corunners:
+                el['names'] = corunners
+            else:
+                del_list.append(el)
+        elif el['type'] == 'task':
+            el['names'] = tasks
+        else:
+            del_list.append(el)
+    for el in del_list:
+        config_json['elements'].remove(el)
+
+
+    dump_json(config_json, config)
```

Comment:
I am mutating the file in place because [here](https://github.com/charlie45000/corunners-example/blob/pgi/to_p2020/scripts/run.power-qoriq-p2020-ds-p.sh#L77) I copy it in the build directory before calling the build script. do you think that it would still be better to use a tempfile (and I don't know if a tempfile remains if the program fails).

An other solution would be to add a security check to know if the original file is in the config_dir and dump it in that directory if not. What do you think of that?